File: exercises/floor03/tests/test_project_loot.py
import importlib.util
import sys
from pathlib import Path

EX = Path(__file__).resolve().parents[1] / "project_loot.py"

spec = importlib.util.spec_from_file_location("project_loot", EX)
mod = importlib.util.module_from_spec(spec)
sys.modules["project_loot"] = mod
spec.loader.exec_module(mod)


def test_roll_loot_boundaries():
    assert mod.roll_loot(1) == "gold coin"
    assert mod.roll_loot(50) == "gold coin"
    assert mod.roll_loot(51) == "torch"
    assert mod.roll_loot(76) == "protein bar"
    assert mod.roll_loot(91) == "steel sword"
    assert mod.roll_loot(100) == "glowstick of the ancients"


def test_add_to_pack():
    assert mod.add_to_pack({}, "torch") == {"torch": 1}
    assert mod.add_to_pack({"torch": 2}, "torch") == {"torch": 3}


def test_open_chest():
    assert mod.open_chest({}, [3, 76, 3]) == {"gold coin": 2, "protein bar": 1}


def test_open_chest_empty_rolls():
    assert mod.open_chest({"torch": 1}, []) == {"torch": 1}


def test_pack_report():
    assert mod.pack_report({"torch": 2, "gold coin": 5}) == "gold coin x5\ntorch x2"
    assert mod.pack_report({}) == "(empty)"

File: exercises/cpp/floor03/tests/test_project_loot.cpp
#include "../../devtest.h"
#include "../project_loot.cpp"

int main() {
    CHECK_EQ(roll_loot(1), std::string("gold coin"), "low roll");
    CHECK_EQ(roll_loot(50), std::string("gold coin"), "boundary 50");
    CHECK_EQ(roll_loot(51), std::string("torch"), "boundary 51");
    CHECK_EQ(roll_loot(76), std::string("protein bar"), "76");
    CHECK_EQ(roll_loot(91), std::string("steel sword"), "91");
    CHECK_EQ(roll_loot(100), std::string("glowstick of the ancients"), "legendary");

    std::map<std::string, int> pack;
    add_to_pack(pack, "torch");
    CHECK_EQ(pack["torch"], 1, "first torch");
    add_to_pack(pack, "torch");
    CHECK_EQ(pack["torch"], 2, "stacks");

    std::map<std::string, int> chest;
    open_chest(chest, {3, 76, 3});
    CHECK_EQ(chest["gold coin"], 2, "two coins");
    CHECK_EQ(chest["protein bar"], 1, "one bar");

    std::map<std::string, int> rpt = {{"torch", 2}, {"gold coin", 5}};
    CHECK_EQ(pack_report(rpt), std::string("gold coin x5\ntorch x2"),
             "alphabetical report (std::map sorts for free)");
    CHECK_EQ(pack_report({}), std::string("(empty)"), "empty pack");
    return dt_summary();
}
